### PR TITLE
Improve the exception handling

### PR DESCRIPTION
**Pull request by Kanapriya**
### Proposed changes in this pull request

Refer - https://github.com/wso2/product-is/issues/6827

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java [hunk 1]
```diff
@@ -139,16 +142,20 @@ private void writeToUserStoreConfigurationFile(Path userStoreConfigFile, UserSto
...
         }
-        throw new IdentityUserStoreMgtException(msg);
+        throw new IdentityUserStoreInternalException(errorCode, msg);
     }
 
     @Override
     public void deleteUserStore(String domain) throws IdentityUserStoreMgtException {
 
-        deleteUserStores(new String[]{domain});
+        if(isValidDomain(domain, false)) {
```
Comment (emswbandara) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java]: formatting issue

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/IdentityUserStoreClientException.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (emswbandara) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/IdentityUserStoreClientException.java]: change license year

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java [hunk 3]
```diff
@@ -426,4 +435,72 @@ private UserStoreDTO getUserStoreDTO(RealmConfiguration secondaryRealmConfigurat
...
+     * To check whether the provided domain name is valid or not.
+     *
+     * @param domainName user store domain name
+     * @param isAdd      boolean value, this will be true when add a new user store
+     * @return true or false
+     * @throws IdentityUserStoreMgtException
+     */
+    private boolean isValidDomain(String domainName, Boolean isAdd) throws IdentityUserStoreMgtException {
+
+        if (domainName == null || "".equals(domainName)) {
```
Comment (emswbandara) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java]: shall we use StringUtils.isEmpty here?

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java [hunk 4]
```diff
@@ -426,4 +435,72 @@ private UserStoreDTO getUserStoreDTO(RealmConfiguration secondaryRealmConfigurat
...
+
+    /**
+     * Get a List of existing domain names.
+     *
+     * @return : list of domain names
+     * @throws IdentityUserStoreMgtException
+     */
+    private ArrayList<String> getDomainNames() throws IdentityUserStoreMgtException {
+
+        ArrayList<String> domains = new ArrayList<String>();
```
Comment (sachiniWettasinghe) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java]: Shall we improve this to `List<String> domains = new ArrayList<>();`

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java [hunk 5]
```diff
@@ -426,4 +436,72 @@ private UserStoreDTO getUserStoreDTO(RealmConfiguration secondaryRealmConfigurat
...
+            throw new IdentityUserStoreClientException(" User store domain name should not be empty.");
+        }
+
+        if (domainName.contains("_")) {
+            throw new IdentityUserStoreClientException(" User store domain name should not contain \"_\".");
+        }
+
+        if (isAdd) {
+            // if add, user store domain name shouldn't already exists
+            if (getDomainNames().contains(domainName)) {
```
Comment (sachiniWettasinghe) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java]: We can improve this logic to` if (isAdd && getDomainNames().contains(domainName))`

File: components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java [hunk 6]
```diff
@@ -426,4 +436,69 @@ private UserStoreDTO getUserStoreDTO(RealmConfiguration secondaryRealmConfigurat
...
+
+    /**
+     * To check whether the provided domain name is valid or not.
+     *
+     * @param domainName user store domain name
+     * @param isAdd      boolean value, this will be true when add a new user store
+     * @return true or false
+     * @throws IdentityUserStoreMgtException
+     */
+    private boolean isValidDomain(String domainName, Boolean isAdd) throws IdentityUserStoreMgtException {
```
Comment (madurangasiriwardena) [inline components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java]: better to have separate methods to check the domain name
- is valid to add
- is existing

Comment (Kanapriya) [reply components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/dao/impl/FileBasedUserStoreDAOImpl.java]: Fix with e9a9db12c05138fc6a0b5a93bfa691091189a8e1

**Pull request merged**